In [2]:
import pandas as pd
import numpy as np
import math

In [3]:
X_train = pd.read_csv('C:\\Users\himab\Documents\spring2020\ML\hw2\hw2-data\Gaussian_process\X_train.csv', header = None)
y_train = pd.read_csv('C:\\Users\himab\Documents\spring2020\ML\hw2\hw2-data\Gaussian_process\y_train.csv', header = None)
X_test = pd.read_csv('C:\\Users\himab\Documents\spring2020\ML\hw2\hw2-data\Gaussian_process\X_test.csv', header = None)
y_test = pd.read_csv('C:\\Users\himab\Documents\spring2020\ML\hw2\hw2-data\Gaussian_process\y_test.csv', header = None)

In [4]:
#Q1
def Rij(list_1,list_2,b):
    a = []
    s = 0
    for i in range(len(list_1)):
        s += pow((list_1[i]-list_2[i]),2)
    return math.exp(-s/b)

print(Rij([1,2,3],[1,2,3],2))
            
        

1.0


In [5]:
X_train = np.array(X_train)
X_test = np.array(X_test)
y_train = np.array(y_train)
y_test = np.array(y_test)


In [6]:
def Kn(X_train,b):
    list1 = []
    for i in range(len(X_train)):
        list2 = []
        for j in range(len(X_train)):
            list2.append(Rij(X_train[i],X_train[j],b))
        list1.append(list2)
    return list1

In [7]:
def KxDn(x,Dn,b):
    list1 = []
    for j in range(len(Dn)):
        list1.append(Rij(x,Dn[j],b))
    return list1

In [8]:
def mu(x,X_train,b,sig_sq, y):
    list1 = []
    for i in range(len(x)):
        a = np.array(KxDn(x[i],X_train,b))
        second_term = (sig_sq * np.identity(len(X_train))) + Kn(X_train,b)
        second_term_inv = np.linalg.inv(second_term)
        mat_mul1 = np.matmul(a,second_term_inv)
        list1.append(np.matmul(mat_mul1,y))
    return(list1)

In [9]:
print(mu(X_test,X_train,3,1.4,y_train))

[array([-5.14517318]), array([-8.311337]), array([7.29135151]), array([-6.27672371]), array([-3.57470024]), array([1.30213042]), array([-6.71537384]), array([-1.95753727]), array([-11.12873278]), array([12.90651609]), array([-7.69770801]), array([12.45772822]), array([-9.872805]), array([-4.37758161]), array([-0.46914401]), array([-7.57201584]), array([-2.87124369]), array([-0.48116405]), array([12.34522361]), array([-4.21010091]), array([-10.57839215]), array([12.8529501]), array([-4.14864859]), array([-5.84861396]), array([-8.19551812]), array([-8.08687842]), array([2.97927992]), array([0.70660872]), array([-2.15113868]), array([-4.33674759]), array([-7.20822386]), array([-4.61102139]), array([2.82643788]), array([-3.52271938]), array([-5.28871767]), array([4.13518781]), array([8.55129911]), array([11.10101647]), array([10.08973691]), array([-4.26571138]), array([2.86095379]), array([-1.90026996])]


In [10]:
#Part (b) #each entry of x_test
b_values = [5,7,9,11,13,15]
sig_sq_values = [.1,.2,.3,.4,.5,.6,.7,.8,.9,1]
means = {}
for b in b_values:
    means[b] = {}
    for sig_sq in sig_sq_values:
        means[b][sig_sq] = mu(X_test,X_train,b,sig_sq,y_train)
means
        

{5: {0.1: [array([-5.10173808]),
   array([-8.27668726]),
   array([7.00111069]),
   array([-6.07445368]),
   array([-3.76972001]),
   array([1.38008315]),
   array([-7.1268909]),
   array([-1.88195054]),
   array([-11.06468038]),
   array([14.2840752]),
   array([-7.50951233]),
   array([14.44245925]),
   array([-9.8916498]),
   array([-3.97763949]),
   array([-0.87857751]),
   array([-9.60470945]),
   array([-2.30670033]),
   array([-0.61945227]),
   array([12.56690827]),
   array([-5.38229761]),
   array([-10.2893002]),
   array([13.67613076]),
   array([-3.86354454]),
   array([-6.58671839]),
   array([-9.8312883]),
   array([-8.3343748]),
   array([3.73527596]),
   array([0.86455116]),
   array([-2.83591735]),
   array([-4.74292338]),
   array([-8.25368113]),
   array([-5.92675241]),
   array([2.16456782]),
   array([-4.0832975]),
   array([-5.33694819]),
   array([3.4515606]),
   array([9.41361318]),
   array([11.74898824]),
   array([9.65897716]),
   array([-4.13303587]),
   arr

In [11]:
def getRMSEvalues(y_test,mu_value):
    val = y_test - mu_value
    mult = val **2
    rmse = math.sqrt(sum(mult)/42)
    return(rmse)
l = {}
rmse_values = {}
for b in b_values:
    l[b] = {}
    rmse_values[b] = {}
    for var in sig_sq_values:
        l[b][var] = mu(X_test,X_train, b, var,y_train)
        rmse = getRMSEvalues(y_test,l[b][var])
        rmse_values[b][var] = rmse


In [12]:
print('RMSE for given variance and b values')
table = pd.DataFrame(rmse_values)
table

RMSE for given variance and b values


5         7         9         11        13        15
0.1  1.966277  1.920164  1.897650  1.890509  1.895850  1.909605
0.2  1.933137  1.904878  1.902521  1.914983  1.935588  1.959551
0.3  1.923422  1.908082  1.917650  1.938851  1.964600  1.990806
0.4  1.922199  1.915904  1.932517  1.957939  1.985504  2.011918
0.5  1.924771  1.924806  1.945702  1.973218  2.001317  2.027373
0.6  1.929215  1.933704  1.957237  1.985767  2.013881  2.039467
0.7  1.934636  1.942256  1.967406  1.996378  2.024313  2.049466
0.8  1.940585  1.950382  1.976494  2.005606  2.033309  2.058107
0.9  1.946822  1.958095  1.984743  2.013838  2.041320  2.065847
1.0  1.953215  1.965440  1.992344  2.021347  2.048644  2.072978

In [13]:
# d = dict(table.min())
# print(min(d, key=d.get)) gives key of minimum value
table.min().min()

1.8905087802491747

The minimum RMSE value here is around 1.89. The minimum RMSE in the first homework was around 2.2. So Gaussian regression seems to have lesser RMSE than ridge regression. 